addons/account/models/account_move.py

```diff
@@ -2581,20 +2581,43 @@ def _is_readonly(self):
     # -------------------------------------------------------------------------
     def _is_eligible_for_early_payment_discount(self, currency, reference_date):
         self.ensure_one()
-        payment_terms = self.line_ids.filtered(lambda line: line.display_type == 'payment_term')
-        return self.currency_id == currency \
-            and self.move_type in self._early_payment_discount_move_types() \
-            and self.invoice_payment_term_id.early_discount \
-            and (
+        return self._is_eligible_for_early_payment_discount_batched(currency, reference_date)[self.id]
+
+    def _is_eligible_for_early_payment_discount_batched(self, currency, reference_date):
+        is_eligible_by_move_id = {move.id: False for move in self}
+        move_types_for_epd = self._early_payment_discount_move_types()
+        eligible_moves = self.filtered(lambda m:
+                                       m.currency_id == currency
+                                       and m.move_type in move_types_for_epd
+                                       and m.invoice_payment_term_id.early_discount
+                                       )
+        if not eligible_moves:
+            return is_eligible_by_move_id
+
+        # A move is not eligible if any of its payment term lines are reconciled
+        unreconciled_pt_lines = self.env['account.move.line'].search([
+            ('move_id', 'in', eligible_moves.ids),
+            ('display_type', '=', 'payment_term'),
+        ]).filtered(lambda line: line.amount_residual != 0)
+
+        pt_lines_by_move = unreconciled_pt_lines.grouped('move_id')
+        for move in eligible_moves:
+            move_payment_terms = pt_lines_by_move.get(move)
+            if not move_payment_terms:
+                continue
```

Comment:
Before: none of the lines are partially reconciled (there is not  a single linked `account.partial.reconcile` linked)
With your change: at least one line is not fully reconciled (there exists at least one line with some residual amount)

This is a change in behavior